src/main.c

```diff
@@ -0,0 +1,257 @@
+      n = rustls_client_session_read_tls(client_session, buf, n);
+      if(n == 0) {
+        fprintf(stderr, "EOF from ClientSession::read_tls\n");
+        // TODO: What to do here?
+        break;
+      }
+      else if(n < 0) {
+        fprintf(stderr, "Error in ClientSession::read_tls\n");
+        return 1;
+      }
+
+      int result = rustls_client_session_process_new_packets(client_session);
+      if(result != CRUSTLS_OK) {
+        fprintf(stderr, "Error in process_new_packets");
+        return 1;
+      }
+
+      /* Read all available bytes from the client_session until EOF.
+       * Note that EOF here indicates "no more bytes until
+       * process_new_packets", not "stream is closed".
+       */
+      for(;;) {
+        memset(buf, 0, sizeof(buf));
+        n = rustls_client_session_read(client_session, buf, sizeof(buf));
+        if(n == 0) {
+          fprintf(stderr, "EOF from ClientSession::read (this is expected)\n");
+          break;
+        }
+        else if(n < 0) {
+          fprintf(stderr, "Error in ClientSession::read\n");
+          return 1;
+        }
+
+        write_all(STDOUT_FILENO, buf, n);
+      }
+    }
+    if(rustls_client_session_wants_write(client_session) &&
+       (events[0].events & EPOLLOUT) > 0) {
+      fprintf(stderr, "ClientSession wants us to write_tls.\n");
+      memset(buf, 0, sizeof(buf));
+      n = rustls_client_session_write_tls(client_session, buf, sizeof(buf));
+      if(n == 0) {
+        fprintf(stderr, "EOF from ClientSession::write_tls\n");
+        return 1;
+      }
+      else if(n < 0) {
+        fprintf(stderr, "Error in ClientSession::write_tls\n");
+        return 1;
+      }
+
+      write_all(sockfd, buf, n);
+    }
+  }
+
+  return 0;
+}
+
+int
+main(int argc, const char **argv)
+{
+  if(argc <= 2) {
+    fprintf(stderr,
+            "usage: %s hostname path\n\n"
+            "Connect to a host via HTTPS on port 443, make a request for the\n"
+            "given path, and emit response to stdout.\n",
+            argv[0]);
+    return 1;
+  }
+  const char *hostname = argv[1];
+  const char *path = argv[2];
+
+  const void *client_config = rustls_client_config_new();
+
+  int sockfd = make_conn(hostname);
+  if(sockfd < 0) {
+    // No perror because make_conn printed error already.
+    return 1;
+  }
+
+  void *client_session = NULL;
+  int result =
+    rustls_client_session_new(client_config, hostname, &client_session);
+  if(result != CRUSTLS_OK) {
+    return 1;
+  }
+
+  int return_code =
+    send_request_and_read_response(sockfd, client_session, hostname, path);
+  rustls_client_session_free(client_session);
+
+  int sockfd2 = make_conn(hostname);
+  if(sockfd2 < 0) {
+    // No perror because make_conn printed error already.
+    return 1;
+  }
+
+  void *client_session2 = NULL;
+  int result2 =
+    rustls_client_session_new(client_config, hostname, &client_session2);
+  if(result2 != CRUSTLS_OK) {
+    return 1;
+  }
+
+  int return_code2 =
```

Comment:
`return_code2` is never checked. Perhaps you want to `return return_code | return_code2;` at the bottom or otherwise explicitly check both.